### PR TITLE
Updates to support ROCm 4.3 in GPU package

### PR DESCRIPTION
**Pull request by rbberger**
**Summary**

Changes how the CMake build of the GPU package operates to support ROCm 4.3.

**Related Issue(s)**

**Author(s)**

@rbberger (Temple U)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

Without this change the `liblammps.so` isn't linked properly. Some shared libraries will not be linked, which only shows up during runtime with `unknown symbol hipInit` errors.

~Currently, there isn't a hipCUB 4.3.x branch. There are several compilation errors with the develop branch.
So as of right now this only works with `HIP_USE_DEVICE_SORTING=OFF`~

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by arghdos** [comment]
Looking at this now -- re: the hipCub stuff, I believe the 4.3 branch is: https://github.com/ROCmSoftwarePlatform/hipCUB/tree/release/rocm-rel-4.3

**Comment by rbberger** [comment]
@arghdos thanks for pointing out that branch. yes indeed, that one works.

**Comment by arghdos** [comment]
@rbberger -- was the `hip_add_*`-> `add_*` the changes needed to get the link to work?  Or is it the `LD_LIBRARY_PATH`.  Wondering if something changed in our cmake that I should ~~beat someone over the head to~~ get documented :)

**Comment by rbberger** [comment]
@arghdos With the old "variant" I'm was getting link errors that showed at runtime (compiling with `BUILD_SHARED_LIBS=on`). Comparing the `hip_get_devices` with `liblammps.so` showed that the shared library wasn't linking to the ROCm libraries:

```
$ ldd ./hip_get_devices 
	linux-vdso.so.1 (0x00007ffc2d3fb000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f4c592d4000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4c592b1000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4c59162000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f4c59157000)
	libopen-pal.so.40 => /lib/x86_64-linux-gnu/libopen-pal.so.40 (0x00007f4c590a9000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f4c58ec5000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4c58cd3000)
	libamdhip64.so.4 => /opt/rocm/lib/libamdhip64.so.4 (0x00007f4c57f1d000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f4c59308000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f4c57f17000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f4c57f12000)
	libhwloc.so.15 => /lib/x86_64-linux-gnu/libhwloc.so.15 (0x00007f4c57ec1000)
	libevent-2.1.so.7 => /lib/x86_64-linux-gnu/libevent-2.1.so.7 (0x00007f4c57e6b000)
	libevent_pthreads-2.1.so.7 => /lib/x86_64-linux-gnu/libevent_pthreads-2.1.so.7 (0x00007f4c57e64000)
	libhsa-runtime64.so.1 => /opt/rocm/lib/libhsa-runtime64.so.1 (0x00007f4c579e6000)
	libnuma.so.1 => /lib/x86_64-linux-gnu/libnuma.so.1 (0x00007f4c579d9000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007f4c579aa000)
	libltdl.so.7 => /lib/x86_64-linux-gnu/libltdl.so.7 (0x00007f4c5799f000)
	libhsakmt.so.1 => /opt/rocm/lib/libhsakmt.so.1 (0x00007f4c57774000)
	libelf.so.1 => /lib/x86_64-linux-gnu/libelf.so.1 (0x00007f4c57758000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f4c5773c000)
$ ldd ./liblammps.so
	linux-vdso.so.1 (0x00007ffc133f5000)
	libopen-pal.so.40 => /lib/x86_64-linux-gnu/libopen-pal.so.40 (0x00007fc1e1a23000)
	libjpeg.so.8 => /lib/x86_64-linux-gnu/libjpeg.so.8 (0x00007fc1e199e000)
	libpng16.so.16 => /lib/x86_64-linux-gnu/libpng16.so.16 (0x00007fc1e1966000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fc1e194a000)
	libfftw3.so.3 => /lib/x86_64-linux-gnu/libfftw3.so.3 (0x00007fc1e1744000)
	libfftw3_omp.so.3 => /lib/x86_64-linux-gnu/libfftw3_omp.so.3 (0x00007fc1e173b000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc1e15ea000)
	libomp.so => /opt/rocm-4.3.0/llvm/lib/libomp.so (0x00007fc1e12f5000)
	libmpi_cxx.so.40 => /lib/x86_64-linux-gnu/libmpi_cxx.so.40 (0x00007fc1e12d7000)
	libmpi.so.40 => /lib/x86_64-linux-gnu/libmpi.so.40 (0x00007fc1e11b2000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fc1e0fd0000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc1e0fb5000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fc1e0f90000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc1e0d9e000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc1e0d98000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fc1e0d93000)
	libhwloc.so.15 => /lib/x86_64-linux-gnu/libhwloc.so.15 (0x00007fc1e0d42000)
	libevent-2.1.so.7 => /lib/x86_64-linux-gnu/libevent-2.1.so.7 (0x00007fc1e0cec000)
	libevent_pthreads-2.1.so.7 => /lib/x86_64-linux-gnu/libevent_pthreads-2.1.so.7 (0x00007fc1e0ce5000)
	libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x00007fc1e0ca3000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc1e2bdc000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fc1e0c98000)
	libopen-rte.so.40 => /lib/x86_64-linux-gnu/libopen-rte.so.40 (0x00007fc1e0bde000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007fc1e0bb1000)
	libltdl.so.7 => /lib/x86_64-linux-gnu/libltdl.so.7 (0x00007fc1e0ba4000)
```
Manually adding the libraries as `target_link_dependencies` resolved the issue.

So naturally, the first thing I did is check the ROCm docs to see if there is something new. Following this page (https://rocmdocs.amd.com/en/latest/Installation_Guide/Using-CMake-with-AMD-ROCm.html) I found out about `find_package(hip)` instead of `find_package(HIP)`. So I gave it a shot and it worked. So it would help to clarify what the difference between the two is.

File: cmake/Modules/Packages/GPU.cmake [hunk 1]
```diff
@@ -219,8 +219,15 @@ elseif(GPU_API STREQUAL "HIP")
           set(HIP_PATH $ENV{HIP_PATH} CACHE PATH "Path to which HIP has been installed")
       endif()
   endif()
-  set(CMAKE_MODULE_PATH "${HIP_PATH}/cmake" ${CMAKE_MODULE_PATH})
-  find_package(HIP REQUIRED)
+  if(NOT DEFINED ROCM_PATH)
+      if(NOT DEFINED ENV{ROCM_PATH})
+          set(ROCM_PATH "/opt/rocm" CACHE PATH "Path to which ROCM has been installed")
```
Comment (akohlmey) [inline cmake/Modules/Packages/GPU.cmake:224]: This is an odd phrase. How about "Path where ROCm is installed" or "Path to ROCm installation"?

**Review by akohlmey — APPROVED** [review]

**Comment by rbberger** [comment]
@arghdos any objections to merge this? We're considering to include this for the stable release, since we're seeing runtime linker errors without it. But it would help to have a final verdict from AMD on this.

**Pull request merged**